python/semantic_kernel/planning/stepwise_planner/stepwise_planner_config.py

```diff
@@ -0,0 +1,25 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+from typing import List, Optional
+
+
+class StepwisePlannerConfig:
```

Comment:
The lists look nice. I especially like the introduction of various callback functions. I have a few questions and additional points. (There might be some misunderstandings on my part due to the lack of examples.)

- I'm not quite sure what `GetSkillFunction` is.
- `GetPromptTemplate`, `PreferTextCompletion`, and `PromptUserConfig` seem like they could already be included in the constructor. Are these moving to `StepwisePlannerConfig` or new concept?
- Personally, since I use `kernel.import_semantic_skill_from_directory` a lot, it would be convenient to expose `SKFunction` for using custom prompt.

Since we're introducing callbacks, it would be nice to include them in other parts a bit more. (I'm curious about what parameters would be included to be used as callbacks and I've provided ideas and opinions with the possibility of manipulating internal flows.). It might increase the complexity of using the planner, but advanced users would probably appreciate it.

1) Providing user callbacks in the `var nextStep = this.ParseResult(actionText);` part could potentially allow for controlling things in the desired way even if errors occur in between. There could be approaches like forcibly creating actions.
<img width="909" alt="image" src="https://github.com/microsoft/semantic-kernel/assets/69464062/98bd2675-7f1b-4d5e-8f19-46d835f24709">  

2) If user callbacks were provided before `return context;`, accessibility for the final result could improve. It seems like it could also solve the issue mentioned in my proposal #2394.  
<img width="775" alt="image" src="https://github.com/microsoft/semantic-kernel/assets/69464062/c9394436-4ed5-4713-b262-a1804fa4a062">

3) Although a clear use case doesn't come to mind immediately, having callback functionality before and after the execution of `InvokeActionAsync` would be beneficial.
<img width="1046" alt="image" src="https://github.com/microsoft/semantic-kernel/assets/69464062/f6f0744d-e948-4efd-bc2f-376ad87979cf">

Reply:
Here's some more context: https://github.com/microsoft/semantic-kernel/pull/1727

>  having callback functionality before and after the execution of InvokeActionAsync would be beneficial.

Agreed! This sounds like part of other efforts right now: https://github.com/microsoft/semantic-kernel/issues/1714

I thought putting them on all the config would be convenient -- what are your thoughts?

Reply:
I agree. It's also good to have everything in the Config.

One additional suggestion is that even though the detailed workings of each planner may be slightly different, the underlying principle is similar (asking questions to the LLM to solve goals and converting them into actions). Because of this, I believe a certain level of abstraction is possible.

However, currently, the implementation of the sequential and step-wise planners is somewhat different (sequential uses `PlanObject`, while step-wise directly calls `InvokeActionAsync`), which might lead to some ambiguity. Nevertheless, in the end, having a unified execution flow would be advantageous in terms of maintenance and user understanding.

I suggest taking this opportunity to abstract appropriately and integrating it into the step-wise planner first, then later refactoring other planners.